### ISSUE TITLE
Level Up

### ISSUE THREAD

**Issue by LuccaHiratsuca** [wooslow/discord-farm-merge-valley-hack#22]
Can we level up using this method? Or does anybody know any way for that?

**Comment by fAntMP** [comment]
> Can we level up using this method? Or does anybody know any way for that?

Yes, you change the expReward breakpoint when merging and you should level up.
Or do you mean the upgrade cards?

**Comment by LennyMcLennington** [comment]
If you set a breakpoint on the line which in the deobfuscated code contains
```js
gameObjectFactory.createById("crate")
```

in current obfuscated code that's this line:
```js
                                this['servi' + 'ces']['visib' + 'leCel' + 'ls'][_0x2ae422(0x344b) + 'lVisi' + _0x2ae422(0x92e)](_0x40721a[_0x2ae422(0x1971) + 'n'], _0x40721a['row'], this['servi' + 'ces']['camer' + 'aServ' + 'ice']['getCa' + 'meraB' + _0x2ae422(0x3382)]()) || this['servi' + 'ces']['camer' + 'aServ' + 'ice']['anima' + 'teCam' + 'eraTo' + 'Cell'](_0x40721a, 0x1),
```

Then spawn a crate, it will trigger the breakpoint. At this point you should write this in the console (numbers will change if the code gets updated):
```js
backendServices = _0x3c4b55.services
worldServices = this.services
behaviors = _0x18c937
```
Then disable breakpoints and continue so that you are able to do things in the game again.

At this point you have access to a lot of useful functionality at any time from the console. Here's some example functions I wrote, you can just paste it into the console after performing the above setup and then run any of the functions whenever you want:

```js
let GridPosition = behaviors.gridPosition,
    gameplayMapScreen = worldServices.world.getAllGameObjects().find(x=>x._blueprintID === "area_cloud").parent.parent.parent.parent.parent,
    rewardSystem = gameplayMapScreen._systems.find(x=>Object.keys(x.__proto__).includes("_processReward")),
    mergeSystem = gameplayMapScreen._systems.find(x=>Object.keys(x).includes("_luckyMergeChance"))

let spawnAtClosestEmptyToCenter = (id) => {
    let object = backendServices.gameObjectFactory.createById(id)
    object.addBehavior(new GridPosition(worldServices.visibleCells.getClosestEmptyToTheCenter()))
    return object
}

let spawnUpgradeCard = (target, rank) => {
    if (!(1 <= rank && rank <= 3)) return;
    let object = spawnAtClosestEmptyToCenter(`upgrade_card_${rank}`)
    object.getBehavior("upgradeCard")._data.targetObjectTreeIngredient = target
    worldServices.world.addGameObject(object)
    return object
}

let spawnCollectable = (target, amount) => {
    let object = spawnAtClosestEmptyToCenter(`gem_1`)
    object.getBehavior("collectable")._data.reward = [{"key": target, "amount": amount}]
    worldServices.world.addGameObject(object)
    return object
}

let giveInventoryItem = (target, amount) => {
    return worldServices.rewardService.giveInventoryReward({
      "reward": {"key": target, "amount": amount},
      "parent": gameplayMapScreen
    })
}

let spawnObject = (target) => {
    let object = spawnAtClosestEmptyToCenter(target)
    worldServices.world.addGameObject(object)
    return object
}

let spawnBubbledObject = (target) => {
    return worldServices.rewardService.giveObjectReward({
      "rewards": [target],
      "container": gameplayMapScreen,
      "animationEndEvent": null,
      "bubblePosition": {"x": 0, "y": -200}
    });
}

// 100 = always lucky merge, 0 = never lucky merge, default is 5
let setLuckyMergeChance = (percentage) => mergeSystem._luckyMergeChance = percentage
```

The function you'd probably be interested in is this one:
```js
let giveInventoryItem = (target, amount) => {
    return worldServices.rewardService.giveInventoryReward({
      "reward": {"key": target, "amount": amount},
      "parent": gameplayMapScreen
    })
}
```

With this set up you can just do `giveInventoryItem("exp", 10000)` or `giveInventoryItem("levels", 1)`. The display won't be updated immediately from this but it is fully functional in every other way, and you can easily refresh it by merging something to gain some xp in a normal way. There's probably a better way to directly give yourself exp but this does work.

**Comment by osu-question** [comment]
> With this set up you can just do `giveInventoryItem("exp", 10000)` or `giveInventoryItem("levels", 1)`. The display won't be updated immediately from this but it is fully functional in every other way, and you can easily refresh it by merging something to gain some xp in a normal way. There's probably a better way to directly give yourself exp but this does work.

Thanks a lot for this, it gave a lot of insight into things to do, but also, I may of done 10k exp thinking it was raw EXP, but somehow I ended up immediately going to level 70, do you think you could help me try to make a function to set your level back down xD, I tried doing negative EXP with the give function but it didn't work

**Comment by hyperwebapps** [comment]
Actually there's also another way that is using these two:
return this['_data']['expRe' + _0x333050(0xc16)] 
return this['_data']['rewar' + 'd']

The first one is the value called for when u craft something and give u experience

**Comment by Xaszanyn** [comment]
> ```js
> backendServices = _0x25af97.services
> worldServices = this.services
> behaviors = _0x5049dc
> ```

They obfuscated again, i could get worldServices with "this._services" but couldn't get backendServices and behaviors.

**Comment by Xaszanyn** [comment]
> ```js
> let GridPosition = behaviors.gridPosition,
>     gameplayMapScreen = worldServices.world.getAllGameObjects().find(x=>x._blueprintID === "area_cloud").parent.parent.parent.parent.parent,
>     rewardSystem = gameplayMapScreen._systems.find(x=>x._processReward),
>     mergeSystem = gameplayMapScreen._systems.find(x=>x._luckyMergeChance),
>     upgradeCardSystem = gameplayMapScreen._systems.find(x=>x._updateUpgradeCardObject)
> ```

I couldn't get any object that has "area_cloud" blueprintID.

**Comment by LennyMcLennington** [comment]
> > ```js
> > let GridPosition = behaviors.gridPosition,
> >     gameplayMapScreen = worldServices.world.getAllGameObjects().find(x=>x._blueprintID === "area_cloud").parent.parent.parent.parent.parent,
> >     rewardSystem = gameplayMapScreen._systems.find(x=>x._processReward),
> >     mergeSystem = gameplayMapScreen._systems.find(x=>x._luckyMergeChance),
> >     upgradeCardSystem = gameplayMapScreen._systems.find(x=>x._updateUpgradeCardObject)
> > ```
> 
> I couldn't get any object that has "area_cloud" blueprintID.

@Xaszanyn it works for me still, but could you try changing "area_cloud" in that to "trainstation" or the blueprint name of some other object that you have placed down in the grid? I'll try finding a more reliable way if that works.

**Comment by bimaindrawans** [comment]
> > > ```js
> > > let GridPosition = behaviors.gridPosition,
> > >     gameplayMapScreen = worldServices.world.getAllGameObjects().find(x=>x._blueprintID === "area_cloud").parent.parent.parent.parent.parent,
> > >     rewardSystem = gameplayMapScreen._systems.find(x=>x._processReward),
> > >     mergeSystem = gameplayMapScreen._systems.find(x=>x._luckyMergeChance),
> > >     upgradeCardSystem = gameplayMapScreen._systems.find(x=>x._updateUpgradeCardObject)
> > > ```
> > 
> > 
> > I couldn't get any object that has "area_cloud" blueprintID.
> 
> @Xaszanyn it works for me still, but could you try changing "area_cloud" in that to "trainstation" or the blueprint name of some other object that you have placed down in the grid? I'll try finding a more reliable way if that works.

do you know how to spawn max item from harvest a plant or animal ?
so for example
i want to spawn gem_5 & cow_4 from harvesting a corn_4

or spawn stone_6 from removing a tree
befor this update i use removeReward but it's seems not working now.

sorry for my bad english but i hope you understand my point.

**Comment by Xaszanyn** [comment]
I unlocked all the fields, that's why it couldn't find any objects, so finding "trainstation" will be safer.

If someone spawned "area_cloud" accidentally like me you can remove only by code. Game will crash eventually when you try to interract with it.

```js
let despawnObject = (column, row) => {
  let object = [...worldServices.world._gameObjects].find(
    (gameObject) =>
      gameObject._behaviors.get("gridPosition").column == column && gameObject._behaviors.get("gridPosition").row == row
  );
  if (object) worldServices.world.removeGameObject(object);
};
```

**Comment by LennyMcLennington** [comment]
> I unlocked all the fields, that's why it couldn't find any objects, so finding "trainstation" will be safer.

@Xaszanyn okay thank you, I edited my original comment with a more reliable method that doesn't require the existence of any specific object being in the map.

**Comment by LennyMcLennington** [comment]
@Xaszanyn BTW, I think this would be a more concise way of writing your function:

```js
let despawnObject = (column, row) => {
  let object = worldServices.mapGrid.getContent(column, row)
  if (object) worldServices.world.removeGameObject(object);
};
```
